iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -653,6 +657,26 @@ public async void OnMessageReceived(Message message)
             }
         }
 
+        private async Task CompleteIncomingMessageAsync(Message message)
+        {
+            try
+            {
+                if (_qosReceivePacketFromService == QualityOfService.AtLeastOnce)
+                {
+                    lock (_syncRoot)
```

Comment:
What if we made this a ConcurrentQueue, so we don't have to lock around it in all cases?